lib/common/ipc_client.c

```diff
@@ -31,6 +31,672 @@
+             * do that unconditionally, because the application might call this
+             * function after a disconnect that happened by other means.
+             */
+            free_on_disconnect = api->free_on_disconnect = true;
+        }
+        pcmk_disconnect_ipc(api); // Frees api if free_on_disconnect is true
+    }
+    if (!free_on_disconnect) {
+        free_daemon_specific_data(api);
+        crm_trace("Freeing IPC API object");
+        free(api);
+    }
+}
+
+/*!
+ * \brief Get the IPC name used with an IPC API connection
+ *
+ * \param[in] api      IPC API connection
+ * \param[in] for_log  If true, return human-friendly name instead of IPC name
+ *
+ * \return IPC API's human-friendly or connection name, or if none is available,
+ *         "Pacemaker" if for_log is true and NULL if for_log is false
+ */
+const char *
+pcmk_ipc_name(pcmk_ipc_api_t *api, bool for_log)
+{
+    if (api == NULL) {
+        return for_log? "Pacemaker" : NULL;
+    }
+    switch (api->server) {
+        case pcmk_ipc_attrd:
+            return for_log? "attribute manager" : NULL /* T_ATTRD */;
+
+        case pcmk_ipc_based:
+            return for_log? "CIB manager" : NULL /* PCMK__SERVER_BASED_RW */;
+
+        case pcmk_ipc_controld:
+            return for_log? "controller" : NULL /* CRM_SYSTEM_CRMD */;
+
+        case pcmk_ipc_execd:
+            return for_log? "executor" : NULL /* CRM_SYSTEM_LRMD */;
+
+        case pcmk_ipc_fenced:
+            return for_log? "fencer" : NULL /* "stonith-ng" */;
+
+        case pcmk_ipc_pacemakerd:
+            return for_log? "launcher" : NULL /* CRM_SYSTEM_MCP */;
+
+        case pcmk_ipc_schedulerd:
+            return for_log? "scheduler" : NULL /* CRM_SYSTEM_PENGINE */;
+
+        default:
+            return for_log? "Pacemaker" : NULL;
+    }
+}
+
+/*!
+ * \brief Check whether an IPC API connection is active
+ *
+ * \param[in] api  IPC API connection
+ *
+ * \return true if IPC is connected, false otherwise
+ */
+bool
+pcmk_ipc_is_connected(pcmk_ipc_api_t *api)
+{
+    return (api != NULL) && crm_ipc_connected(api->ipc);
+}
+
+/*!
+ * \internal
+ * \brief Call the daemon-specific API's dispatch function
+ *
+ * Perform daemon-specific handling of IPC reply dispatch. It is the daemon
+ * method's responsibility to call the client's registered event callback, as
+ * well as allocate and free any event data.
+ *
+ * \param[in] api  IPC API connection
+ */
+static void
+call_api_dispatch(pcmk_ipc_api_t *api, xmlNode *message)
+{
+    crm_log_xml_trace(message, "ipc-received");
+    if ((api->cmds != NULL) && (api->cmds->dispatch != NULL)) {
+        api->cmds->dispatch(api, message);
+    }
+}
+
+/*!
+ * \internal
+ * \brief Dispatch data read from IPC source
+ *
+ * \param[in] buffer     Data read from IPC
+ * \param[in] length     Number of bytes of data in buffer (ignored)
+ * \param[in] user_data  IPC object
+ *
+ * \return Always 0 (meaning connection is still required)
+ *
```

Comment:
If the return value is always 0, it doesn't seem like there's a point in having a return value.

Reply:
The function signature is defined by struct ipc_client_callbacks's dispatch() member, which in turn is defined by mainloop_io_t's dispatch_fn_ipc() member. That one uses a return value of 0 to tell glib to keep the connection alive after dispatching, and a negative value to tell glib to close the connection after dispatching. Here, we never want to close it, so we always return 0.

Reply:
Ah, okay.  That makes sense.